uefi: fs: Implement exists

Ayush1325 opened the pull request:
Also adds the initial file abstractions.

The file opening algorithm is inspired from UEFI shell. It starts by classifying if the Path is Shell mapping, text representation of device path protocol, or a relative path and converts into an absolute text representation of device path protocol.

After that, it queries all handles supporting
EFI_SIMPLE_FILE_SYSTEM_PROTOCOL and opens the volume that matches the device path protocol prefix (similar to Windows drive). After that, it opens the file in the volume using the remaining pat.

It also introduces OwnedDevicePath and BorrowedDevicePath abstractions to allow working with the base UEFI and Shell device paths efficiently.

DevicePath in UEFI behaves like an a group of nodes laied out in the memory contiguously and thus can be modeled using iterators.

This is an effort to break the original PR (https://github.com/rust-lang/rust/pull/129700) into much smaller chunks for faster upstreaming.

jhpratt commented on library/std/src/sys/pal/uefi/fs.rs:
```diff
@@ -342,3 +347,185 @@ pub fn canonicalize(_p: &Path) -> io::Result<PathBuf> {
+            return Ok(ans);
+        }
+
+        // Absoulte Device Path
```
nit: typo (in a few places)

jhpratt reviewed (commented):
ping @dvdhrm and @nicholasbishop as UEFI target maintainers. r=me with approval from either of them and the typo fixed.

nicholasbishop commented:
There are still quite a few changes in this PR. For better review, could we break it up into some smaller pieces? For example, maybe we could start with a PR that just does some of the device-path helper changes.

Ayush1325 commented:
> There are still quite a few changes in this PR. For better review, could we break it up into some smaller pieces? For example, maybe we could start with a PR that just does some of the device-path helper changes.

Would it be fine to have some dead code with the cfg attribute? I can add the device psth abstraction and the file open stuff separately in that case.

Ayush1325 commented:
> There are still quite a few changes in this PR. For better review, could we break it up into some smaller pieces? For example, maybe we could start with a PR that just does some of the device-path helper changes.

I have moved the changes to the old DevicePath to a separate PR: https://github.com/rust-lang/rust/pull/135393

It only contains stuff for OwnedDevicePath, but it should reduce all the noise due to renaming and refactoring from this PR.

bors commented:
:umbrella: The latest upstream changes (presumably #135465) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Ayush1325 commented:
@nicholasbishop Does this PR seem manageable now?

nicholasbishop commented on library/std/src/sys/pal/uefi/fs.rs:
```diff
@@ -342,3 +347,122 @@ pub fn canonicalize(_p: &Path) -> io::Result<PathBuf> {
+            vol.open(&mut path_remaining, open_mode, attr)
+        }
+
+        fn open_volume_from_device_path(
```
Add a docstring, and include information about what's being returned. Include an example of the device path, e.g. 
```
For example, given "PciRoot(0x0)/Pci(0x1,0x1)/Ata(Secondary,Slave,0x0)/\abc\run.efi", this will open the volume ""PciRoot(0x0)/Pci(0x1,0x1)/Ata(Secondary,Slave,0x0)" and return the remaining file path "\abc\run.efi".
```

Also, are there restrictions on `path`, like needing to be absolute? If so state that as well.

nicholasbishop commented on library/std/src/sys/pal/uefi/fs.rs:
```diff
@@ -342,3 +347,122 @@ pub fn canonicalize(_p: &Path) -> io::Result<PathBuf> {
+        }
+    }
+
+    fn path_best_match<'a>(
```
Add docstring

nicholasbishop commented on library/std/src/sys/pal/uefi/fs.rs:
```diff
@@ -342,3 +347,122 @@ pub fn canonicalize(_p: &Path) -> io::Result<PathBuf> {
+
+    fn path_best_match<'a>(
+        source: &helpers::BorrowedDevicePath<'a>,
+        target: &helpers::BorrowedDevicePath<'a>,
```
nit: `<'_>` and drop the named lifetime?

nicholasbishop commented on library/std/src/sys/pal/uefi/helpers.rs:
```diff
@@ -330,6 +389,99 @@ impl<'a> crate::fmt::Debug for BorrowedDevicePath<'a> {
     }
 }
 
+pub(crate) struct DevicePathIterator<'a>(Option<DevicePathNode<'a>>);
```
Can you move the addition of the iterator to a separate PR?

Ayush1325 replied:
Sure. I have limited bandwidth rn, so will do it in a few weeks.

Ayush1325 replied:
I have created the PR: https://github.com/rust-lang/rust/pull/137424

nicholasbishop reviewed (commented):
Thanks for your patience, was out sick for a bit. Left a few comments, but I think it would still be helpful to break this down a little further, perhaps starting with `DevicePathIterator`.

bors commented:
:umbrella: The latest upstream changes (presumably #138279) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Ayush1325 committed:
uefi: fs: Implement exists

Also adds the initial file abstractions.

The file opening algorithm is inspired from UEFI shell. It starts by
classifying if the Path is Shell mapping, text representation of device
path protocol, or a relative path and converts into an absolute text
representation of device path protocol.

After that, it queries all handles supporting
EFI_SIMPLE_FILE_SYSTEM_PROTOCOL and opens the volume that matches the
device path protocol prefix (similar to Windows drive). After that, it
opens the file in the volume using the remaining pat.

It also introduces OwnedDevicePath and BorrowedDevicePath abstractions
to allow working with the base UEFI and Shell device paths efficiently.

DevicePath in UEFI behaves like an a group of nodes laied out in the
memory contiguously and thus can be modeled using iterators.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 commented:
@nicholasbishop This is ready for review now.

nicholasbishop commented on library/std/src/sys/fs/uefi.rs:
```diff
@@ -317,3 +322,141 @@ pub fn canonicalize(_p: &Path) -> io::Result<PathBuf> {
+        /// 2. Path relative to the volume.
+        ///
+        /// For example, given "PciRoot(0x0)/Pci(0x1,0x1)/Ata(Secondary,Slave,0x0)/\abc\run.efi",
+        /// this will open the volume ""PciRoot(0x0)/Pci(0x1,0x1)/Ata(Secondary,Slave,0x0)"
```
```suggestion
        /// this will open the volume "PciRoot(0x0)/Pci(0x1,0x1)/Ata(Secondary,Slave,0x0)"
```

nicholasbishop commented on library/std/src/sys/fs/uefi.rs:
```diff
@@ -317,3 +322,141 @@ pub fn canonicalize(_p: &Path) -> io::Result<PathBuf> {
+                simple_file_system::PROTOCOL_GUID,
+            )?;
+
+            let mut file_protocol: MaybeUninit<*mut file::Protocol> = MaybeUninit::uninit();
```
Is `MaybeUninit` needed here? I think this could be simplified to `let mut file_protocol = null_mut();` and pass in `&mut file_protocol` as the parameter. Then you can drop the `unsafe` `assume_init` below. (I don't think the current code is wrong, just trying to reduce `unsafe` code.)

nicholasbishop commented on library/std/src/sys/fs/uefi.rs:
```diff
@@ -317,3 +322,141 @@ pub fn canonicalize(_p: &Path) -> io::Result<PathBuf> {
+
+        fn open(&self, path: &mut [u16], open_mode: u64, attr: u64) -> io::Result<Self> {
+            let file_ptr = self.0.as_ptr();
+            let mut file_opened: MaybeUninit<*mut file::Protocol> = MaybeUninit::uninit();
```
Ditto on dropping `MaybeUninit`.

nicholasbishop reviewed (approved):
lgtm

jhpratt commented:
@bors r=jhpratt,nicholasbishop

bors commented:
:pushpin: Commit 2e70cfc04d29ec499e076746d2588d8e7a4e0fe6 has been approved by `jhpratt,nicholasbishop`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jhpratt,nicholasbishop 2e70cfc04d29ec499e076746d2588d8e7a4e0fe6 -->
<!-- homu: {"type":"Approved","sha":"2e70cfc04d29ec499e076746d2588d8e7a4e0fe6","approver":"jhpratt,nicholasbishop","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 2e70cfc04d29ec499e076746d2588d8e7a4e0fe6 with merge 75530e9f72a1990ed2305e16fd51d02f47048f12...
<!-- homu: {"type":"BuildStarted","head_sha":"2e70cfc04d29ec499e076746d2588d8e7a4e0fe6","merge_sha":"75530e9f72a1990ed2305e16fd51d02f47048f12"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13919376527/job/38960322969)
Approved by: jhpratt,nicholasbishop
Pushing 75530e9f72a1990ed2305e16fd51d02f47048f12 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jhpratt,nicholasbishop","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13919376527/job/38960322969"},"merge_sha":"75530e9f72a1990ed2305e16fd51d02f47048f12"} -->

rust-timer commented:
Finished benchmarking commit (75530e9f72a1990ed2305e16fd51d02f47048f12): [comparison URL](https://perf.rust-lang.org/compare.html?start=259fdb521200c9abba547302fc2c826479ef26b2&end=75530e9f72a1990ed2305e16fd51d02f47048f12&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=259fdb521200c9abba547302fc2c826479ef26b2&end=75530e9f72a1990ed2305e16fd51d02f47048f12&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=259fdb521200c9abba547302fc2c826479ef26b2&end=75530e9f72a1990ed2305e16fd51d02f47048f12&stat=max-rss)
<details>
<summary>Results (primary -2.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.9% | [-2.9%, -2.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.9% | [-2.9%, -2.9%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=259fdb521200c9abba547302fc2c826479ef26b2&end=75530e9f72a1990ed2305e16fd51d02f47048f12&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=259fdb521200c9abba547302fc2c826479ef26b2&end=75530e9f72a1990ed2305e16fd51d02f47048f12&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 776.409s -> 776.359s (-0.01%)
**Artifact size**: 365.11 MiB -> 365.12 MiB (0.00%)
